src/util/locking/FileSystemResourceLocker.ts

```diff
@@ -0,0 +1,160 @@
+  retries: 0,
+};
+
+const defaultUnlockOptions: UnlockOptions = {
+  /** Resolve symlinks using realpath, defaults to true (note that if true, the file must exist previously) */
+  realpath: false,
+};
+
+const attemptDefaults: Required<AttemptSettings> = { retryCount: -1, retryDelay: 50, retryJitter: 30 };
+
+/**
+ * Argument interface of the FileSystemResourceLocker constructor.
+ */
+interface FileSystemResourceLockerArgs {
+  /** The rootPath of the filesystem */
+  rootFilePath?: string;
+  /** The path to the directory where locks will be stored (appended to rootFilePath) */
+  lockDirectory?: string;
+  /** Custom settings concerning retrying locks */
+  attemptSettings?: AttemptSettings;
+}
+
+function isCodedError(err: unknown): err is { code: string } & Error {
+  return typeof err === 'object' && err !== null && 'code' in err;
+}
+
+/**
+ * A resource locker making use of the [proper-lockfile](https://www.npmjs.com/package/proper-lockfile) library.
+ * Note that no locks are kept in memory, thus this is considered thread- and process-safe.
+ *
+ * This **proper-lockfile** library has its own retry mechanism for the operations, since a lock/unlock call will
+ * either resolve successfully or reject immediately with the causing error. The retry function of the library
+ * however will be ignored and replaced by our own LockUtils' {@link retryFunctionUntil} function.
+ *
+ * This allows us to reuse the AttemptSettings semantics similar to the {@link RedisLocker} class.
+ */
+export class FileSystemResourceLocker implements ResourceLocker, Finalizable {
+  protected readonly logger = getLoggerFor(this);
+  private readonly attemptSettings: Required<AttemptSettings>;
+  /** Folder that stores the locks */
+  private readonly lockFolder: string;
+
+  /**
+   * Create a new FileSystemResourceLocker
+   * @param rootFilePath - The rootPath of the filesystem _[default is the current dir `./`]_
+   * @param lockDirectory - The path to the directory where locks will be stored (appended to rootFilePath)
+                            _[default is `/.internal/locks`]_
+   * @param attemptSettings - Custom settings concerning retrying locks
+   */
+  public constructor(args: FileSystemResourceLockerArgs = {}) {
+    const { rootFilePath, lockDirectory, attemptSettings } = args;
+    this.attemptSettings = { ...attemptDefaults, ...attemptSettings };
+    this.lockFolder = joinFilePath(rootFilePath ?? './', lockDirectory ?? '/.internal/locks');
+    ensureDirSync(this.lockFolder);
+  }
+
+  /**
+   * Wrapper function for all un/lock operations. Any errors coming from the `fn()` will be swallowed.
+   * Only `ENOTACQUIRED` errors wills be thrown (trying to release lock that didn't exist).
+   * @param fn - The function reference to swallow errors from.
+   */
+  private swallowErrors(fn: () => Promise<VoidFunctionOrVoid>): () => Promise<BooleanOrVoid> {
```

Comment:
```suggestion
  private swallowErrors(fn: () => Promise<VoidFunctionOrVoid>): () => Promise<BooleanOrUndefined> {
```
A void return type of a function means the output of the function should be ignored. `undefined` specifically means that the function returns the `undefined` value.

And perhaps also add a comment in the tsdoc explaining that undefined is returned because that is what `tryFunctionUntil` expects if a retry needs to happen.

`unknown` would also have been fine instead of `VoidFunctionOrVoid` since we don't actually care about the output. I just don't like `any` because it hides all type errors if you accidentally use it, `unknown` gives an error when trying to use it. But keeping `VoidFunctionOrVoid` is also fine 😄